web/src/main/java/org/cbioportal/web/util/AlterationCountUtil.java

```diff
@@ -0,0 +1,109 @@
+import org.cbioportal.service.AlterationCountService;
+import org.cbioportal.service.GenePanelService;
+import org.cbioportal.web.parameter.AlterationCountFilter;
+import org.cbioportal.web.parameter.Projection;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+import java.util.*;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+@Component
+public class AlterationCountUtil {
+    @Autowired
+    private AlterationCountService alterationCountService;
+    @Autowired
+    private GenePanelService genePanelService;
+    
+    public Pair<List<AlterationCountByGene>, Long> getAlterationCounts(
+        AlterationCountFilter alterationCountFilter,
+        AlterationCountType alterationCountType,
+        List<Integer> entrezGeneIds
+    ) {
+        // select all genes if no list of entrez gene ids provided
+        Select<Integer> geneSelect = (entrezGeneIds == null || entrezGeneIds.isEmpty()) ?
+            Select.all(): Select.byValues(entrezGeneIds);
+        List<MolecularProfileCaseIdentifier> molecularProfileCaseIdentifiers =
+            alterationCountFilter.getMolecularProfileCaseIdentifiers();
+        boolean includeFrequency = true;
+        boolean includeMissingAlterationsFromGenePanel = false;
+        AlterationFilter alterationFilter = alterationCountFilter.getAlterationFilter();
+
+         return alterationCountType.equals(AlterationCountType.SAMPLE) ?
+            alterationCountService.getSampleAlterationCounts(
+                molecularProfileCaseIdentifiers,
+                geneSelect,
+                includeFrequency,
+                includeMissingAlterationsFromGenePanel,
+                alterationFilter
+            ) :
+            alterationCountService.getPatientAlterationCounts(
+                molecularProfileCaseIdentifiers,
+                geneSelect,
+                includeFrequency,
+                includeMissingAlterationsFromGenePanel,
+                alterationFilter
+            );
+    }
+    
+    public Set<GenePanelToGene> getProfiledGenes(
+        AlterationCountFilter alterationCountFilter,
+        AlterationCountType alterationCountType,
+        List<Integer> entrezGeneIds
+    ) {
+        List<MolecularProfileCaseIdentifier> molecularProfileCaseIdentifiers =
+            alterationCountFilter.getMolecularProfileCaseIdentifiers();
+        Set<GenePanelToGene> profiledGenes;
+        List<GenePanelData> genePanelDataList = alterationCountType.equals(AlterationCountType.SAMPLE) ?
+            genePanelService.fetchGenePanelDataInMultipleMolecularProfiles(molecularProfileCaseIdentifiers) :
+            genePanelService.fetchGenePanelDataInMultipleMolecularProfilesByPatientIds(molecularProfileCaseIdentifiers);
+
+        if (genePanelDataList
+            .stream()
+            .anyMatch(
+                genePanelData -> genePanelData.getProfiled() && genePanelData.getGenePanelId() == null
```

Comment:
Remind me again - why do we allow gene panel data objects that don't have a gene panel? What does the object even represent at that point?